src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php

```diff
@@ -97,12 +103,75 @@ private function completeDefinition($id, Definition $definition)
             $this->container->addResource(static::createResourceForClass($reflectionClass));
         }
 
-        if (!$constructor = $reflectionClass->getConstructor()) {
-            return;
+        $methodsCalled = array();
+        foreach ($definition->getMethodCalls() as $methodCall) {
+            $methodsCalled[$methodCall[0]] = true;
         }
 
-        $arguments = $definition->getArguments();
-        foreach ($constructor->getParameters() as $index => $parameter) {
+        foreach ($this->getMethodsToAutowire($id, $reflectionClass, $autowiredMethods) as $reflectionMethod) {
+            if (!isset($methodsCalled[$reflectionMethod->name])) {
+                $this->autowireMethod($id, $definition, $reflectionMethod);
+            }
+        }
+    }
+
+    /**
+     * Gets the list of methods to autowire.
+     *
+     * @param string           $id
+     * @param \ReflectionClass $reflectionClass
+     * @param string[]         $autowired
+     *
+     * @return \ReflectionMethod[]
+     */
+    private function getMethodsToAutowire($id, \ReflectionClass $reflectionClass, array $autowired)
+    {
+        $found = array();
+        $regexList = array();
+        foreach ($autowired as $pattern) {
+            $regexList[] = '/^'.str_replace('\*', '.*', preg_quote($pattern, '/')).'$/i';
+        }
+
+        foreach ($reflectionClass->getMethods(\ReflectionMethod::IS_PUBLIC) as $reflectionMethod) {
+            if ($reflectionMethod->isStatic()) {
+                continue;
+            }
+
+            foreach ($regexList as $k => $regex) {
+                if (preg_match($regex, $reflectionMethod->name)) {
+                    $found[] = $autowired[$k];
+                    yield $reflectionMethod;
+
+                    continue 2;
+                }
+            }
+        }
+
+        if ($notFound = array_diff($autowired, $found)) {
+            $compiler = $this->container->getCompiler();
+            $compiler->addLogMessage($compiler->getLoggingFormatter()->formatUnusedAutowiringPatterns($this, $id, $notFound));
+        }
+    }
+
+    /**
+     * Autowires the constructor or a setter.
+     *
+     * @param string            $id
+     * @param Definition        $definition
+     * @param \ReflectionMethod $reflectionMethod
+     *
+     * @throws RuntimeException
+     */
+    private function autowireMethod($id, Definition $definition, \ReflectionMethod $reflectionMethod)
+    {
+        if ($isConstructor = $reflectionMethod->isConstructor()) {
+            $arguments = $definition->getArguments();
+        } else {
+            $arguments = array();
+        }
+
+        $addMethodCall = false;
```

Comment:
Could you add a small comment here? I don't think it's easy to understand what contains this variable without looking deeply the code.
Maybe `// Whether the method should be added to the definition as a call`.

At the same time, I don't remember if it's written somewhere in the code what's this pass behavior (e.g. only autowire methods when at least one if their arguments is not to its default value), it could be worth adding it if not.